lib/libc/cheri/strfcap.3

```diff
@@ -0,0 +1,173 @@
+.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+.\" SUCH DAMAGE.
+.\"
+.Dd February 11, 2021
+.Dt STRFCAP 3
+.Os
+.Sh NAME
+.Nm strfcap
+.Nd format decoded CHERI capability
+.Sh LIBRARY
+.Lb libc
+.Sh SYNOPSIS
+.In cheri/cheric.h
+.Ft size_t
+.Fo strfcap
+.Fa "char * restrict buf"
+.Fa "size_t maxsize"
+.Fa "const char * restrict format"
+.Fa "uintcap_t cap"
+.Fc
+.Sh DESCRIPTION
+The
+.Fn strfcap
+function formats information from decoding
+.Fa cap
+into the buffer
+.Fa buf
+according to the string pointed to by
+.Fa format .
+.Pp
+The
+.Fa format
+string consists of zero or more conversion specifications and
+ordinary characters.
+All ordinary characters are copied directly into the buffer.
+A conversion specification consists of a percent sign
+.Dq Ql %
+and one other character.
+.Pp
+No more than
+.Fa maxsize
+characters will be placed into the array.
+If the total number of resulting characters, including the terminating
+NUL character, is not more than
+.Fa maxsize ,
+.Fn strftime
+returns the number of characters in the array, not counting the
+terminating NUL.
+Otherwise, zero is returned and the buffer contents are indeterminate.
+.Pp
+The conversion specifications are copied to the buffer after expansion
+as follows:-
+.Bl -tag -width "xxxx"
+.It Cm \&%a
+is replaced by address of the capability.
+.It Cm \&%A
+is replaced by a textual representation of capablity attributes
+enclosed in parentheses.
+Zero or more of the follow attributes are included:
+.Pp
+.Bl -tag -compact -width invalid
+.It invalid
+Capability's tag is clear.
+.It sentry
+Capability is a sealed entry.
+.It sealed
+Capability is sealed but not a sealed entry.
+.El
+.Pp
+If no attributes are present, the field is omitted.
+.It Cm \&%b
+is replaced by the base address of the capablity.
+.It Cm \&%c
+if
+.Fa cap
+is NULL-derived this is equivalent to
+.Ql 0x%a .
+If
+.Fa cap
+is not NULL-derived it is equivalent to
+.Ql 0x%a [%P,0x%b-0x%t]%? %A .
+This is equivalent to
+the
+.Ql %#p
+format of
+.Xr printf 3 .
+.It Cm \&%p
+is replaced by the raw permissions in hexadecimal format.
+.It Cm \&%P
+is replaced by an abbreviated textual representation of permissions
+listing zero or more of the following characters:
+.Pp
+.Bl -tag -compact -width W
+.It r
+LOAD permission
+.It w
+STORE permission
+.It x
+EXECUTE permission
+.It R
+LOAD_CAP permission
+.It W
+STORE_CAP permission
+.It E
+EXECUTIVE permission (Morello only)
+.El
```

Comment:
Seal, Unseal, CID, CCall, ASR, ...?

Reply:
For common use we're eliding less common permissions (though @bsdjhb suggests X for ASR in https://github.com/CTSRD-CHERI/cheri-c-programming/wiki/Displaying-Capabilities). Another option or adding a `#` modifier with a more complete set might be appropriate.

A related question is: what to do about user-permissions when some are reserved and some are not.